### PR TITLE
new layout (separable palette) tidies

### PR DESCRIPTION
**Pull request by ahgittin**
#84 introduces the separable palette/config panel

this enhances that so that the palette gets the right size, and also fixes extension points and makes it so the "add (member spec)" usage of the panel respects those extension points

minor tidies to dropdowns and other style points also

### CODE REVIEW

**Comment by ahgittin** [comment]
retest this please

File: ui-modules/blueprint-composer/app/components/catalog-selector/catalog-selector.directive.js [hunk 1]
```diff
@@ -108,15 +144,21 @@ export function catalogSelectorSortFilter($filter) {
...
-function controller($scope, $element, $q, $uibModal, $log, $templateCache, paletteApi, paletteDragAndDropService, iconGenerator, composerOverrides) {
+function controller($scope, $element, $timeout, $q, $uibModal, $log, $templateCache, paletteApi, paletteDragAndDropService, iconGenerator, composerOverrides) {
+    this.$timeout = $timeout;
+
+    $scope.viewModes = PALETTE_VIEW_MODES;
+    $scope.viewOrders = ['name', 'type', 'id'];
+    
+    if (!$scope.state) $scope.state = {};
+    if (!$scope.state.viewMode) $scope.state.viewMode = PALETTE_VIEW_MODES.normal;
+    if (!$scope.state.currentOrder) $scope.state.currentOrder = 'name';
```
Comment (tbouron) [inline ui-modules/blueprint-composer/app/components/catalog-selector/catalog-selector.directive.js]: Could you use curly brackets for the `if` statement please? That's to be consistent with the rest of the codebase and also make it easier to read

File: ui-modules/blueprint-composer/app/components/catalog-selector/catalog-selector.directive.js [hunk 2]
```diff
@@ -127,15 +169,15 @@ function controller($scope, $element, $q, $uibModal, $log, $templateCache, palet
...
+    
     $scope.isLoading = true;
 
     $scope.$watch('search', () => {
         $scope.freeFormTile = {
             symbolicName: $scope.search,
             name: $scope.search,
             displayName: $scope.search,
-            supertypes: [$scope.family.superType]
+            supertypes: ($scope.family ? [ $scope.family.superType ] : []),
```
Comment (tbouron) [inline ui-modules/blueprint-composer/app/components/catalog-selector/catalog-selector.directive.js]: `$scope.family` is marked as required so this should not be needed. Callers of the `catalog-selector` directive should always pass it

File: ui-modules/blueprint-composer/app/components/providers/palette-service.provider.js [hunk 3]
```diff
@@ -37,7 +37,9 @@ export function paletteServiceProvider() {
             sections[id] = section;
         },
         deleteSection(id) {
+            let old = sections[id];
             delete sections[id];
+            return old;
```
Comment (tbouron) [inline ui-modules/blueprint-composer/app/components/providers/palette-service.provider.js]: I'm curious, why returning `old`?

File: ui-modules/blueprint-composer/app/views/main/graphical/edit/add/add.js [hunk 4]
```diff
@@ -51,8 +52,11 @@ export function GraphicalEditAddController($scope, $filter, $state, $stateParams
             break;
     }
 
-    $scope.catalogItemsPerPage = 24;
-
+    if (!$scope.familiesToShow) $scope.familiesToShow = [ $scope.family ];
```
Comment (tbouron) [inline ui-modules/blueprint-composer/app/views/main/graphical/edit/add/add.js]: Could you use curly brackets for the `if` statement please? That's to be consistent with the rest of the codebase and also make it easier to read

File: ui-modules/blueprint-composer/app/views/main/graphical/edit/add/add.js [hunk 5]
```diff
@@ -27,17 +27,18 @@ export const graphicalEditAddState = {
...
 
-export function GraphicalEditAddController($scope, $filter, $state, $stateParams, blueprintService) {
+export function GraphicalEditAddController($scope, $filter, $state, $stateParams, blueprintService, paletteService) {
     switch ($stateParams.family) {
         case EntityFamily.ENTITY.id.toLowerCase():
             $scope.family = EntityFamily.ENTITY;
             break;
         case EntityFamily.SPEC.id.toLowerCase():
             $scope.family = EntityFamily.SPEC;
+            $scope.familiesToShow = [ EntityFamily.ENTITY, EntityFamily.SPEC ];
```
Comment (tbouron) [inline ui-modules/blueprint-composer/app/views/main/graphical/edit/add/add.js]: Seems to be used only in the JS code so would be better to be a local var (i.e. avoiding adding stuff to the `$scope` as a general rule of thumb)

File: ui-modules/blueprint-composer/app/components/catalog-selector/catalog-selector.directive.js [hunk 6]
```diff
@@ -243,4 +285,6 @@ function controller($scope, $element, $q, $uibModal, $log, $templateCache, palet
 
     // allow downstream to configure this controller and/or scope
     (composerOverrides.configurePaletteController || function() {})(this, $scope, $element);
+    
+    $scope.paletteItemClasses = "col-xs-3";
```
Comment (tbouron) [inline ui-modules/blueprint-composer/app/components/catalog-selector/catalog-selector.directive.js]: This doesn't seems to be used so can be removed

File: ui-modules/blueprint-composer/app/components/catalog-selector/catalog-selector.directive.js [hunk 7]
```diff
@@ -20,7 +20,14 @@ import angular from 'angular';
 import {EntityFamily} from '../util/model/entity.model';
 import template from './catalog-selector.template.html';
 
-const ITEMS_PER_PAGE = 30;
+const MIN_ROWS_PER_PAGE = 4;
+
+const PALETTE_VIEW_MODES = {
```
[the remark this answers is not in the document]
Comment (ahgittin) [reply ui-modules/blueprint-composer/app/components/catalog-selector/catalog-selector.directive.js]: not convinced.  could be changed when useful.

File: ui-modules/blueprint-composer/app/components/catalog-selector/catalog-selector.directive.js (hunk 2 above)
Comment (ahgittin) [reply ui-modules/blueprint-composer/app/components/catalog-selector/catalog-selector.directive.js]: can be passed in as null however

File: ui-modules/blueprint-composer/app/components/providers/palette-service.provider.js (hunk 3 above)
Comment (ahgittin) [reply ui-modules/blueprint-composer/app/components/providers/palette-service.provider.js]: if you want to change order a common pattern would be to delete it but put it to a local var so you could then add it in the right place.  same idea as `Map.remove`

File: ui-modules/blueprint-composer/app/views/main/graphical/edit/add/add.js (hunk 5 above)
Comment (ahgittin) [reply ui-modules/blueprint-composer/app/views/main/graphical/edit/add/add.js]: no, it's used in the html

File: ui-modules/blueprint-composer/app/components/catalog-selector/catalog-selector.directive.js (hunk 6 above)
Comment (ahgittin) [reply ui-modules/blueprint-composer/app/components/catalog-selector/catalog-selector.directive.js]: yes, removed in #94

**Comment by ahgittin** [comment]
all addressed as part of #93

File: ui-modules/blueprint-composer/app/views/main/graphical/edit/add/add.js (hunk 4 above)
Comment (tbouron) [reply ui-modules/blueprint-composer/app/views/main/graphical/edit/add/add.js]: I cannot find any reference if this on templates

File: ui-modules/blueprint-composer/app/views/main/graphical/edit/add/add.js (hunk 5 above)
Comment (tbouron) [reply ui-modules/blueprint-composer/app/views/main/graphical/edit/add/add.js]: I cannot find any reference if this on templates

Comment (ahgittin) [reply ui-modules/blueprint-composer/app/views/main/graphical/edit/add/add.js]: https://github.com/apache/brooklyn-ui/pull/93/files#diff-ed969f9a529823af9b9282a89cc5ebc1R31

**Comment by tbouron** [comment]
Thanks @ahgittin, merging

**Pull request merged**